nvcc version do not support C++20 std::source_location completely

python -m minisgl --model "../../Qwen/Qwen3-0.6B"

`
[2025-12-20|08:57:37] INFO     Parsed arguments:
ServerArgs(model_path='../../Qwen/Qwen3-0.6B', tp_info=DistributedInfo(rank=0, size=1), dtype=torch.bfloat16, max_running_req=256, attention_backend='auto', cuda_graph_bs=None, cuda_graph_max_bs=None, page_size=1, memory_ratio=0.9, distributed_timeout=60.0, use_dummy_weight=False, use_pynccl=True, max_seq_len_override=None, num_page_override=None, max_extend_tokens=8192, cache_type='radix', offline_mode=False, _unique_suffix='.pid=2657', server_host='127.0.0.1', server_port=1919, num_tokenizer=0, silent_output=False)
[2025-12-20|08:57:40|initializer] INFO     Tokenize server 0 is ready
[2025-12-20|08:57:41|core|rank=0] INFO     Free memory before loading model: 78.82 GiB
[2025-12-20|08:57:41|core|rank=0] INFO     Allocating 650122 pages for KV cache, K + V = 69.44 GiB
[2025-12-20|08:57:41|core|rank=0] INFO     Auto-selected attention backend: fi
[2025-12-20|08:57:41|core|rank=0] INFO     Free memory after initialization: 7.71 GiB
[2025-12-20|08:57:41|core|rank=0] INFO     Start capturing CUDA graphs with sizes: [160, 152, 144, 136, 128, 120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 4, 2, 1]
[2025-12-20|08:57:41|core|rank=0] INFO     Free GPU memory before capturing CUDA graphs: 7.64 GiB
Process minisgl-TP0-scheduler:
Traceback (most recent call last):
  File "/root/miniconda3/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/root/miniconda3/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/root/autodl-tmp/mini-sglang/python/minisgl/server/launch.py", line 21, in _run_scheduler
    scheduler = Scheduler(args)
                ^^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/scheduler/scheduler.py", line 84, in __init__
    self.engine = Engine(config)
                  ^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/engine/engine.py", line 101, in __init__
    self.graph_runner = GraphRunner(
                        ^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/engine/graph.py", line 96, in __init__
    self.logits[:] = model.forward()
                     ^^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/models/qwen3.py", line 82, in forward
    output = self.model.forward(ctx.batch.input_ids)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/models/qwen3.py", line 61, in forward
    x = self.embed_tokens.forward(input_ids)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/layers/embedding.py", line 35, in forward
    y = indexing(
        ^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/kernel/index.py", line 48, in indexing
    module = _jit_index_module(element_size, num_splits=num_splits)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/kernel/index.py", line 23, in _jit_index_module
    return load_jit(
           ^^^^^^^^^
  File "/root/autodl-tmp/mini-sglang/python/minisgl/kernel/utils.py", line 120, in load_jit
    return load_inline(
           ^^^^^^^^^^^^
  File "/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/cpp/extension.py", line 892, in load_inline
    build_inline(
  File "/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/cpp/extension.py", line 740, in build_inline
    return _build_impl(
           ^^^^^^^^^^^^
  File "/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/cpp/extension.py", line 540, in _build_impl
    build_ninja(str(build_dir))
  File "/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/cpp/extension.py", line 412, in build_ninja
    raise RuntimeError("\n".join(msg))
RuntimeError: ninja exited with status 2
stdout:
[1/2] /usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output cuda_0.o.d -Xcompiler -fPIC -std=c++17 -O2 -gencode=arch=compute_80,code=sm_80 -std=c++20 -O3 --expt-relaxed-constexpr -I/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/include -I/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/include -I/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include -c /root/.cache/tvm-ffi/minisgl__index_2048_4_128_1_false_f634a34728e4e029/cuda.cu -o cuda_0.o
FAILED: [code=2] cuda_0.o 
/usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output cuda_0.o.d -Xcompiler -fPIC -std=c++17 -O2 -gencode=arch=compute_80,code=sm_80 -std=c++20 -O3 --expt-relaxed-constexpr -I/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/include -I/root/miniconda3/lib/python3.12/site-packages/tvm_ffi/include -I/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include -c /root/.cache/tvm-ffi/minisgl__index_2048_4_128_1_false_f634a34728e4e029/cuda.cu -o cuda_0.o
nvcc warning : incompatible redefinition for option 'std', the last value of this option was used
nvcc warning : incompatible redefinition for option 'optimize', the last value of this option was used
/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include/minisgl/utils.h(44): error: call to consteval function "std::source_location::current" did not produce a valid constant expression
                                       std::source_location::current()) {
                                                                    ^
/usr/include/c++/11/source_location(59): note #2703-D: cannot call non-constexpr function "__builtin_source_location" (declared implicitly)
      current(__builtin_ret_type __p = __builtin_source_location()) noexcept
                                       ^

/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include/minisgl/utils.h(55): error: call to consteval function "std::source_location::current" did not produce a valid constant expression
        std::source_location location = std::source_location::current()) {
                                                                     ^
/usr/include/c++/11/source_location(59): note #2703-D: cannot call non-constexpr function "__builtin_source_location" (declared implicitly)
      current(__builtin_ret_type __p = __builtin_source_location()) noexcept
                                       ^

/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include/minisgl/tensor.h(415): error: call to consteval function "std::source_location::current" did not produce a valid constant expression
                Loc_t loc = Loc_t::current()) const && -> const TensorMatcher && {
                                          ^
/usr/include/c++/11/source_location(59): note #2703-D: cannot call non-constexpr function "__builtin_source_location" (declared implicitly)
      current(__builtin_ret_type __p = __builtin_source_location()) noexcept
                                       ^

/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include/minisgl/utils.cuh(65): error: call to consteval function "std::source_location::current" did not produce a valid constant expression
             std::source_location location = std::source_location::current())
                                                                          ^
/usr/include/c++/11/source_location(59): note #2703-D: cannot call non-constexpr function "__builtin_source_location" (declared implicitly)
      current(__builtin_ret_type __p = __builtin_source_location()) noexcept
                                       ^

/root/autodl-tmp/mini-sglang/python/minisgl/kernel/csrc/include/minisgl/utils.cuh(74): error: call to consteval function "std::source_location::current" did not produce a valid constant expression
  CUDA_CHECK(std::source_location location = std::source_location::current())
                                                                          ^
/usr/include/c++/11/source_location(59): note #2703-D: cannot call non-constexpr function "__builtin_source_location" (declared implicitly)
      current(__builtin_ret_type __p = __builtin_source_location()) noexcept
                                       ^

5 errors detected in the compilation of "/root/.cache/tvm-ffi/minisgl__index_2048_4_128_1_false_f634a34728e4e029/cuda.cu".
ninja: build stopped: subcommand failed.
`

`
gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
`

`
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Mon_Apr__3_17:16:06_PDT_2023
Cuda compilation tools, release 12.1, V12.1.105
Build cuda_12.1.r12.1/compiler.32688072_0
`

`
nvidia-smi
Sat Dec 20 09:01:01 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.65.06              Driver Version: 580.65.06      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A800 80GB PCIe          On  |   00000000:4F:00.0 Off |                  Off |
| N/A   31C    P0             44W /  300W |       0MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A800 80GB PCIe          On  |   00000000:D5:00.0 Off |                  Off |
| N/A   32C    P0             41W /  300W |       0MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

the command nvcc test.cu -o test && ./test can work normally.
`

--- a/python/minisgl/kernel/csrc/include/minisgl/tensor.h
+++ b/python/minisgl/kernel/csrc/include/minisgl/tensor.h
@@ -32,6 +32,12 @@ struct SizeRef;
 struct DTypeRef;
 struct DeviceRef;
 
+inline constexpr auto kAnyDeviceID = -1;
+inline constexpr auto kAnySize = static_cast<int64_t>(-1);
+inline constexpr auto kNullSize = static_cast<int64_t>(0);
+inline constexpr auto kNullDType = static_cast<DLDataTypeCode>(18u);
+inline constexpr auto kNullDevice = static_cast<DLDeviceType>(-1);
+
 template <typename T> struct dtype_trait {};
 
 template <std::integral T> struct dtype_trait<T> {
@@ -49,23 +55,18 @@ template <std::floating_point T> struct dtype_trait<T> {
                  .lanes = 1};
 };
 
-inline constexpr auto kAnyDeviceID = -1;
-inline constexpr auto kAnySize = static_cast<int64_t>(-1);
-inline constexpr auto kNullSize = static_cast<int64_t>(0);
-inline constexpr auto kNullDType = static_cast<DLDataTypeCode>(18u);
-inline constexpr auto kNullDevice = static_cast<DLDeviceType>(-1);
+template <DLDeviceType Code> struct device_trait {
+  inline static constexpr auto value =
+      DLDevice{.device_type = Code, .device_id = kAnyDeviceID};
+};
 
 template <typename... Ts>
 inline constexpr auto kDTypeList =
     std::array<DLDataType, sizeof...(Ts)>{dtype_trait<Ts>::value...};
 
-template <auto Code>
-inline constexpr auto _kOneDevice = DLDevice{
-    .device_type = static_cast<DLDeviceType>(Code), .device_id = kAnyDeviceID};
-
-template <auto... Codes>
+template <DLDeviceType... Codes>
 inline constexpr auto kDeviceList =
-    std::array<DLDevice, sizeof...(Codes)>{_kOneDevice<Codes>...};
+    std::array<DLDevice, sizeof...(Codes)>{device_trait<Codes>::value...};
 
 template <typename T> struct PrintAbleSpan {
   explicit PrintAbleSpan(std::span<const T> data) : data(data) {}

--- a/python/minisgl/kernel/csrc/include/minisgl/utils.h
+++ b/python/minisgl/kernel/csrc/include/minisgl/utils.h
@@ -1,10 +1,37 @@
 #pragma once
 
+// ref:
+// https://forums.developer.nvidia.com/t/c-20s-source-location-compilation-error-when-using-nvcc-12-1/258026/3
+#ifdef __CUDACC__
+#pragma push_macro("__cpp_consteval")
+#pragma push_macro("_NODISCARD")
+#pragma push_macro("__builtin_LINE")
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wbuiltin-macro-redefined"
+#define __cpp_consteval 201811L
+#pragma clang diagnostic pop
+
+#ifdef _NODISCARD
+#undef _NODISCARD
+#define _NODISCARD
+#endif
+
+#define consteval constexpr
+
+#include <source_location>
+
+#undef consteval
+#pragma pop_macro("__cpp_consteval")
+#pragma pop_macro("_NODISCARD")
+#else
+#include <source_location>
+#endif
+
 #include <dlpack/dlpack.h>
 
 #include <concepts>
 #include <ostream>
-#include <source_location>
 #include <sstream>
 #include <utility>
 

--- a/python/minisgl/kernel/csrc/include/minisgl/warp.cuh
+++ b/python/minisgl/kernel/csrc/include/minisgl/warp.cuh
@@ -4,8 +4,6 @@
 #include <sys/cdefs.h>
 
 #include <cstddef>
-#include <cstdint>
-#include <type_traits>
 
 namespace device::warp {
 
@@ -37,40 +35,6 @@ inline constexpr auto resolve_unit_size(std::size_t x) -> std::size_t {
 template <std::size_t kBytes, std::size_t kUnit>
 using mem_package_t = decltype(get_mem_package<kUnit>());
 
-template <typename T, std::size_t N> struct uint_vec {
-  T data[N];
-};
-
-__always_inline __device__ auto load_nc(const uint1 *__restrict__ src)
-    -> uint1 {
-  uint32_t tmp;
-  asm volatile("ld.global.cs.b32 %0,[%1];" : "=r"(tmp) : "l"(src));
-  return uint1{tmp};
-}
-
-__always_inline __device__ auto load_nc(const uint2 *__restrict__ src)
-    -> uint2 {
-  uint32_t tmp0, tmp1;
-  asm volatile("ld.global.cs.v2.b32 {%0,%1},[%2];"
-               : "=r"(tmp0), "=r"(tmp1)
-               : "l"(src));
-  return uint2{tmp0, tmp1};
-}
-
-__always_inline __device__ void store_nc(uint1 *__restrict__ dst,
-                                         const uint1 &value) {
-  uint32_t tmp = value.x;
-  asm volatile("st.global.cs.b32 [%0],%1;" ::"l"(dst), "r"(tmp));
-}
-
-__always_inline __device__ void store_nc(uint2 *__restrict__ dst,
-                                         const uint2 &value) {
-  uint32_t tmp0 = value.x;
-  uint32_t tmp1 = value.y;
-  asm volatile("st.global.cs.v2.b32 [%0],{%1,%2};" ::"l"(dst), "r"(tmp0),
-               "r"(tmp1));
-}
-
 } // namespace details
 
 template <std::size_t kBytes,
@@ -91,52 +55,6 @@ __always_inline __device__ void copy(void *__restrict__ dst,
   for (std::size_t i = 0; i < kLoopCount; ++i) {
     const auto j = i * kWarpThreads + lane_id;
     dst_packed[j] = src_packed[j];
-  }
-}
-
-template <std::size_t kBytes,
-          std::size_t kUnit = details::resolve_unit_size(kBytes)>
-__always_inline __device__ auto load_vec(const void *__restrict__ src) {
-  using Package = details::mem_package_t<kBytes, kUnit>;
-  constexpr auto kBytesPerLoop = sizeof(Package) * kWarpThreads;
-  constexpr auto kLoopCount = kBytes / kBytesPerLoop;
-  using vec_t = details::uint_vec<Package, kLoopCount>;
-  static_assert(kBytes % kBytesPerLoop == 0,
-                "kBytes must be multiple of 128 bytes");
-
-  const auto src_packed = static_cast<const Package *>(src);
-  const auto lane_id = threadIdx.x % kWarpThreads;
-  vec_t vec;
-
-#pragma unroll kLoopCount
-  for (std::size_t i = 0; i < kLoopCount; ++i) {
-    const auto j = i * kWarpThreads + lane_id;
-    // vec.data[i] = src_packed[j];
-    vec.data[i] = details::load_nc(src_packed + j);
-  }
-  return vec;
-}
-
-template <std::size_t kBytes,
-          std::size_t kUnit = details::resolve_unit_size(kBytes), typename T>
-__always_inline __device__ void store_vec(void *__restrict__ dst,
-                                          const T &vec) {
-  using Package = details::mem_package_t<kBytes, kUnit>;
-  constexpr auto kBytesPerLoop = sizeof(Package) * kWarpThreads;
-  constexpr auto kLoopCount = kBytes / kBytesPerLoop;
-  using vec_t = details::uint_vec<Package, kLoopCount>;
-  static_assert(kBytes % kBytesPerLoop == 0,
-                "kBytes must be multiple of 128 bytes");
-  static_assert(std::is_same_v<T, vec_t>, "store_vec: src type mismatch");
-
-  const auto dst_packed = static_cast<Package *>(dst);
-  const auto lane_id = threadIdx.x % kWarpThreads;
-
-#pragma unroll kLoopCount
-  for (std::size_t i = 0; i < kLoopCount; ++i) {
-    const auto j = i * kWarpThreads + lane_id;
-    // dst_packed[j] = vec.data[i];
-    details::store_nc(dst_packed + j, vec.data[i]);
   }
 }
 

--- a/python/minisgl/kernel/csrc/jit/index.cu
+++ b/python/minisgl/kernel/csrc/jit/index.cu
@@ -88,7 +88,6 @@ __global__ __launch_bounds__(kNumThreads, kMaxOccupancy) void //
                                        (warp_id % kNumSplits) * kSizePerWarp);
       warp::copy<kSizePerWarp>(dst, src);
     } else {
-      // memset the warp to zero, using uint4 package
       warp::reset<kSizePerWarp>(dst);
     }
   }

